File: c++/Solution98.cpp
//DFS(BST) O(V) O(H)
class Solution98 {
public:
    bool isValidBST(TreeNode* root) { 
        return dfs(root, nullptr, nullptr); 
    }

    bool dfs(TreeNode* root, TreeNode* min_node, TreeNode* max_node) {
        if (!root) return true;
        if (min_node && min_node->val >= root->val) return false; // null時表示無限制
        if (max_node && root->val >= max_node->val) return false;
        return dfs(root->left, min_node, root) && dfs(root->right, root, max_node);
    }
};


//DFS(BST) O(V) O(H)
class Solution98_2 {
public:
    bool isValidBST(TreeNode* root) { 
        return dfs(root, LONG_MIN, LONG_MAX); // 用long否則會有極值問題, root-val=INT_MAX or INT_MIN 會誤判
    }

    bool dfs(TreeNode* root, long min_val, long max_val) {
        if (!root) return true;
        if (min_val >= root->val || root->val >= max_val) return false;
        return dfs(root->left, min_val, root->val) && dfs(root->right, root->val, max_val);
    }
};

File: java/Solution98.java
package leetCode.java;

import java.util.*;

//DFS(BST) O(V) O(H)
class Solution98 {
    public boolean isValidBST(TreeNode root) {
         // 用null避免邊界問題, 如果用數字判斷就要用long, 因為node.val可能是MAX_VALUE or MIN_VALUE
        return dfs(root, null, null);
    }

    public boolean dfs(TreeNode root, TreeNode minNode, TreeNode maxNode) {
        if (root == null) {
            return true;
        }
        if (minNode != null && minNode.val >= root.val) { // min or max null 代表沒限制
            return false;
        }
        if (maxNode != null && maxNode.val <= root.val) {
            return false;
        }
        return dfs(root.left, minNode, root) && dfs(root.right, root, maxNode);
    }
}


//DFS(BST) O(V) O(H)
class Solution98_2 {
    public boolean isValidBST(TreeNode root) {
        long minVal = Long.MIN_VALUE; // 用long避免極值問題, 否則 left=Integer.MIN_VALUE會比較不出來
        Deque<TreeNode> stack = new ArrayDeque<>();
        while (!stack.isEmpty() || root != null) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
            root = stack.pop();
            if (root.val <= minVal) { // BST嚴格比較, 等於代表不符合
                return false;
            }
            minVal = root.val;
            root = root.right;
        }
        return true;
    }
}